pkg/kubelet/volume/persistent_claim/persistent_claim.go

```diff
@@ -0,0 +1,149 @@
+)
+
+func (plugin *persistentClaimPlugin) Init(host volume.Host) {
+	plugin.host = host
+}
+
+func (plugin *persistentClaimPlugin) Name() string {
+	return persistentClaimPluginName
+}
+
+func (plugin *persistentClaimPlugin) CanSupport(spec *api.Volume) bool {
+	return spec.PersistentVolumeClaimVolumeSource != nil
+}
+
+func (plugin *persistentClaimPlugin) NewBuilder(spec *api.Volume, podRef *api.ObjectReference) (volume.Builder, error) {
+
+	glog.V(5).Infof("Creating builder for %+v\n", spec)
+
+	claimName := spec.PersistentVolumeClaimVolumeSource.ClaimName
+	volNamespace := podRef.Namespace
+	claim, err := plugin.host.GetKubeClient().PersistentVolumeClaims(volNamespace).Get(claimName)
+
+	glog.V(5).Infof("Using claim: %+v\n", claim)
+
+	if err != nil {
+		glog.Errorf("Error finding claim in namespace  PersistentVolume by ClaimName: %+v\n", spec.PersistentVolumeClaimVolumeSource.ClaimName)
+		return nil, err
+	}
+
+	if claim != nil {
+		pv, err := plugin.host.GetKubeClient().PersistentVolumes().Get(claim.Status.VolumeRef.Name)
+
+		if err != nil {
+			glog.Errorf("Error finding bound PersistentVolume by ClaimName: %+v\n", spec.PersistentVolumeClaimVolumeSource.ClaimName)
+			return nil, err
+		}
+
+		glog.V(5).Infof("Using persistent volume: %+v\n", pv)
+		volPlugin := getVolumePlugin(pv.Spec)
+
+		if volPlugin == nil {
+			glog.Errorf("No plugin found for PersistentVolume: %+v\n", pv)
+			return nil, err
+		}
+
+		glog.V(5).Infof("Using plugin: %s\n", volPlugin.Name())
+		volSource, err := getVolumeSource((pv.Spec.PersistentVolumeSource))
+
+		if err != nil {
+			glog.Errorf("Error getting VolumeSource for %+v: \n", pv.Spec.PersistentVolumeSource)
+			return nil, err
+		}
+
+		glog.V(5).Infof("Using volume source: %+v\n", volSource)
+		wrapper := &api.Volume{
+			Name:         spec.Name,
+			VolumeSource: volSource,
+		}
+
+		volPlugin.Init(plugin.host)
+		builder, err := volPlugin.NewBuilder(wrapper, podRef)
+
+		glog.V(5).Infof("Returning builder: %s\n", builder.GetPath())
+
+		if err != nil {
+			glog.Errorf("Error creating builder for ClaimName: %+v\n", spec.PersistentVolumeClaimVolumeSource.ClaimName)
+			return nil, err
+		}
+
+		if builder != nil {
+			return builder, nil
+		}
+	}
+
+	return nil, fmt.Errorf("No builder found for volume %+v\n", spec)
+}
+
+func (plugin *persistentClaimPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return nil, fmt.Errorf("This will never be called directly.  Volumes backing a PV have their own cleaners and Kubelet reconciles volumes in its syncLoop.")
+}
+
+func getVolumeSource(pv api.PersistentVolumeSource) (api.VolumeSource, error) {
+	if pv.HostPath != nil {
+		return api.VolumeSource{
+			HostPath: pv.HostPath,
+		}, nil
+	} else if pv.GCEPersistentDisk != nil {
+		return api.VolumeSource{
+			GCEPersistentDisk: pv.GCEPersistentDisk,
+		}, nil
+	}
+	return api.VolumeSource{}, fmt.Errorf("No set member found on %+v:", pv)
+}
```

Comment:
@thockin This method is currently necessary because Kubelet only works with VolumeSources.  An interface that covers VolumeSource and PersistentVolumeSource would be necessary to remove this.  Another consequence is that the VS would need to be present in both Volume and PersistentVolume for this wrapping to work.